### PR TITLE
Multiscale Vision Transformers

### PR DESCRIPTION
**Pull request by Amapocho**
#14

### CODE REVIEW

**Review by abhi-glitchhg — CHANGES_REQUESTED** [review]
Could you please use Feedforward class from `encoder/nn` as MLP Instead of 'common/mlp'?

Also you can import DropPath from timm library, use this import statement - `from timm.models.layers import DropPath`, 

Also if Pull Resquest is still in progress, you can change it to Draft.

**Comment by Amapocho** [comment]
> Could you please use Feedforward class from `encoder/nn` as MLP Instead of 'common/mlp'?
> 
> Also you can import DropPath from timm library, use this import statement - `from timm.models.layers import DropPath`,
> 
> Also if Pull Resquest is still in progress, you can change it to Draft.

Made the necessary changes

File: vformer/models/classification/multiscale.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
+import math
+import torch
+import torch.nn as nn
+
+from vformer.common import BaseClassificationModel
+from vformer.encoder.embedding.patch_multiscale import PatchEmbed
+from vformer.decoder.mlp import MLPDecoder
+from vformer.encoder.multiscale import MultiScaleBlock
+
```
Comment (abhi-glitchhg) [inline vformer/models/classification/multiscale.py:9]: I would suggest to use relative imports

File: vformer/encoder/multiscale.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
+import torch
+from torch import nn
+
+from timm.models.layers import DropPath
+from .nn import FeedForward as Mlp
+from ..attention import MultiScaleAttention
+
```
Comment (abhi-glitchhg) [inline vformer/encoder/multiscale.py:7]:
```suggestion
from ..attention import MultiScaleAttention
from ..utils import ENCODER_REGISTRY

@ENCODER_REGISTRY.register()
```

File: vformer/attention/multiscale.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        tensor = norm(tensor)
+    # Assert tensor_dim in [3, 4]
+    if tensor_dim == 4:
+        pass
+    else:  #  tensor_dim == 3:
+        tensor = tensor.squeeze(1)
+    return tensor, thw_shape
+
+
+class MultiScaleAttention(nn.Module):
```
Comment (abhi-glitchhg) [inline vformer/attention/multiscale.py:56]:
```suggestion
@ATTENTION_REGISTRY.register()
class MultiScaleAttention(nn.Module):
```

File: vformer/encoder/embedding/patch_multiscale.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+class PatchEmbed(nn.Module):
```
Comment (abhi-glitchhg) [inline vformer/encoder/embedding/patch_multiscale.py:1]:
```suggestion
import torch.nn as nn

class PatchEmbed(nn.Module):
```

File: vformer/encoder/multiscale.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        self.pool_skip = (
+            nn.MaxPool3d(kernel_skip, stride_skip, padding_skip, ceil_mode=False)
+            if len(kernel_skip) > 0
+            else None
+        )
+
+    def forward(self, x, thw_shape):
+        x_block, thw_shape_new = self.attn(self.norm1(x), thw_shape)
+        x_res, _ = attention_pool(
```
Comment (abhi-glitchhg) [inline vformer/encoder/multiscale.py:119]: here You need to import attention_pool  from the embedding module.

File: tests/test_attention.py [hunk 6]
```diff
@@ -51,6 +51,22 @@ def test_CrossAttention():
     assert out.shape == test_tensor1.shape
     del attention
 
+def test_MultiScaleAttention():
+
+    test_tensor1 = torch.randn(96,8,56,56)
+    test_tensor2 = torch.randn(768,8,14,14)
+    thw  = [2,2,2]
```
Comment (abhi-glitchhg) [inline tests/test_attention.py:58]: I did one more review round; 
i guess you will need to create 5 dimensional input tensors (Batch,Channels,Time,Height,width) according to the authors code

File: vformer/models/classification/multiscale.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+             norm_stem = False
+             norm_layer = partial(nn.LayerNorm, eps=1e-6)
+             patch_kernel = (3, 7, 7)
+             patch_stride = (2, 4, 4)
+             patch_padding = (1, 3, 3)
+             DIM_MUL: [[1, 2.0], [3, 2.0], [14, 2.0]]
+             HEAD_MUL: [[1, 2.0], [3, 2.0], [14, 2.0]]
+             POOL_KVQ_KERNEL: [3, 3, 3]
+             POOL_KV_STRIDE_ADAPTIVE: [1, 8, 8]
+             POOL_Q_STRIDE: [[1, 1, 2, 2], [3, 1, 2, 2], [14, 1, 2, 2]]
```
Comment (abhi-glitchhg) [inline vformer/models/classification/multiscale.py:45]:
```suggestion
             patch_kernel = (3, 7, 7),
             patch_padding = (1, 3, 3),
             DIM_MUL=[[1, 2.0], [3, 2.0], [14, 2.0]],
             HEAD_MUL= [[1, 2.0], [3, 2.0], [14, 2.0]],
             POOL_KVQ_KERNEL= [3, 3, 3],
             POOL_KV_STRIDE_ADAPTIVE=[1, 8, 8],
             POOL_Q_STRIDE= [[1, 1, 2, 2], [3, 1, 2, 2], [14, 1, 2, 2]],
```

File: vformer/models/classification/multiscale.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+             in_chans = 3,
+             use_2d_patch = False,
+             patch_stride = [2,4,4],
+             num_classes = 400
+             embed_dim = 96
+             num_heads = 1
+             mlp_ratio = 4.0
+             qkv_bias = True
+             drop_rate = 0.0
+             depth = 16
```
Comment (abhi-glitchhg) [inline vformer/models/classification/multiscale.py:31]:
```suggestion
             num_classes = 400,
             embed_dim = 96,
             num_heads = 1,
             mlp_ratio = 4.0,
             qkv_bias = True,
             drop_rate = 0.0,
             depth = 16,
```

File: vformer/models/classification/multiscale.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
+import math
+import torch
+import torch.nn as nn
+
```
Comment (abhi-glitchhg) [inline vformer/models/classification/multiscale.py:4]:
```suggestion
import torch.nn as nn
from functools import partial
```

File: vformer/models/classification/multiscale.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        pool_kv = [[] for i in range(depth)]
+        stride_q = [[] for i in range(depth)]
+        stride_kv = [[] for i in range(depth)]
+
+        for i in range(len(POOL_Q_STRIDE)):
+            stride_q[POOL_Q_STRIDE[i][0]] = POOL_Q_STRIDE[i][
+                1:
+            ]
+            if cfg.MVIT.POOL_KVQ_KERNEL is not None:
+                pool_q[POOL_Q_STRIDE[i][0]] = POOL_KVQ_KERNEL
```
Comment (abhi-glitchhg) [inline vformer/models/classification/multiscale.py:121]:
```suggestion
            ]
            if POOL_KVQ_KERNEL is not None:
                pool_q[POOL_Q_STRIDE[i][0]] = POOL_KVQ_KERNEL
```

File: vformer/models/classification/multiscale.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+                POOL_KV_STRIDE.append([i] + _stride_kv)
+
+        for i in range(len(POOL_KV_STRIDE)):
+            stride_kv[POOL_KV_STRIDE[i][0]] = POOL_KV_STRIDE[
+                i
+            ][1:]
+            if POOL_KVQ_KERNEL is not None:
+                pool_kv[
+                    POOL_KV_STRIDE[i][0]
+                ] = KVQ_KERNEL
```
Comment (abhi-glitchhg) [inline vformer/models/classification/multiscale.py:146]: `KVQ_KERNEL`  this variable is not defined .

File: vformer/models/classification/multiscale.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+                kernel_q=pool_q[i] if len(pool_q) > i else [],
+                kernel_kv=pool_kv[i] if len(pool_kv) > i else [],
+                stride_q=stride_q[i] if len(stride_q) > i else [],
+                stride_kv=stride_kv[i] if len(stride_kv) > i else [],
+                mode=mode,
+                has_cls_embed=self.cls_embed_on,
+                pool_first=pool_first,
+            )
+            if cfg.MODEL.ACT_CHECKPOINT:
+                attention_block = checkpoint_wrapper(attention_block)
```
Comment (abhi-glitchhg) [inline vformer/models/classification/multiscale.py:183]:
```suggestion
            )
            if ACT_CHECKPOINT:
                attention_block = checkpoint_wrapper(attention_block)
```

File: vformer/models/classification/multiscale.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+
+        embed_dim = dim_out
+        self.norm = norm_layer(embed_dim)
+
+        self.head = head_helper.TransformerBasicHead(
+            embed_dim,
+            num_classes,
+            dropout_rate=self.drop_rate,
+            act_func=cfg.MODEL.HEAD_ACT,
+        )
```
Comment (abhi-glitchhg) [inline vformer/models/classification/multiscale.py:194]:
```suggestion
            num_classes,
            dropout_rate=self.drop_rate,
            act_func=HEAD_ACT,
        )
```

**Review by abhi-glitchhg — CHANGES_REQUESTED** [review]

**Pull request closed without merging**